components/script/dom/xmlhttprequest.rs

```diff
@@ -679,28 +679,29 @@ impl XMLHttpRequestMethods for XMLHttpRequest {
             None => None,
         };
 
-        let mut request = RequestBuilder::new(self.request_url.borrow().clone().unwrap())
-            .method(self.request_method.borrow().clone())
-            .headers((*self.request_headers.borrow()).clone())
-            .unsafe_request(true)
-            // XXXManishearth figure out how to avoid this clone
-            .body(extracted_or_serialized.map(|e| e.into_net_request_body().0))
-            // XXXManishearth actually "subresource", but it doesn't exist
-            // https://github.com/whatwg/xhr/issues/71
-            .destination(Destination::None)
-            .synchronous(self.sync.get())
-            .mode(RequestMode::CorsMode)
-            .use_cors_preflight(has_handlers)
-            .credentials_mode(credentials_mode)
-            .use_url_credentials(use_url_credentials)
-            .origin(self.global().origin().immutable().clone())
-            .referrer(
-                self.referrer_url
-                    .clone()
-                    .map(|referrer_url| Referrer::ReferrerUrl(referrer_url)),
-            )
-            .referrer_policy(self.referrer_policy.clone())
-            .pipeline_id(Some(self.global().pipeline_id()));
+        let mut request = RequestBuilder::new(
+            self.request_url.borrow().clone().unwrap(),
+            self.referrer_url
+                .clone()
+                .map(|url| Referrer::ReferrerUrl(url))
+                .unwrap_or_else(|| self.global().get_referrer()),
```

Comment:
Let's remove the Option around referrer_url and update [this code](https://github.com/servo/servo/blob/70af0bf14286cadc8f5b6e666131ba1c106331c1/components/script/dom/xmlhttprequest.rs#L171-L176) to use global.get_referrer directly.

Reply:
I wasn't sure if I could change this struct since it is a dom struct and (in my understanding) can get passed to JS. Is this partially true and some fields should not be edited?

Reply:
The contents of the struct are never exposed directly to JS. JS calls methods in the XMLHttpRequestMethods trait, so we can change the struct in any way we want that allows those methods to continue returning expected values.